KAFKA-3854: Fix issues with new consumer's subsequent regex (pattern) subscriptions

vahidhashemian opened the pull request:
This patch fixes two issues:
1. Subsequent regex subscriptions fail with the new consumer.
2. Subsequent regex subscriptions would not immediately refresh metadata to change the subscription of the new consumer and trigger a rebalance.

The final note on the JIRA stating that a later created topic that matches a consumer's subscription pattern would not be assigned to the consumer upon creation seems to be as designed. A repeat 
`subscribe()` to the same pattern or some wait time until the next automatic metadata refresh would handle that case.

An integration test was also added to verify these issues are fixed with this PR.

vahidhashemian committed:
KAFKA-3854: Fix issues with new consumer's subsequent regex (pattern) subscriptions

This patch fixes two issues:
1. Subsequent regex subscriptions fail with the new consumer.
2. Subsequent regex subscriptions would not actually refresh metadata and change the subscription of the new consumer nor they would trigger a rebalance.

The final note on the JIRA stating that a later created topic that matches a consumer's subscription pattern would not be assigned to the consumer upon creation seems to be as designed. A repeat subscribe() to the same pattern would be needed to handle that case.
Unit tests for regex subscriptions will be handled in KAFKA-3897.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -47,6 +47,13 @@
+    };
+
+    /* the type of subscription */
+    private SubscriptionType subscriptionType;
```
Do we need to initialize this to NONE?

hachikuji replied:
Oh, never mind. I see it below

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -77,6 +84,13 @@
     private static final String SUBSCRIPTION_EXCEPTION_MESSAGE =
         "Subscription to topics, partitions and pattern are mutually exclusive";
 
+    private void confirmSubscriptionType(SubscriptionType type) {
```
Nitpick: the use of 'confirm' in the name makes it sound like this is just a check, but it actually mutates internal state. Maybe `setSubscriptionType` would be a little more accurate?

vahidhashemian replied:
I was having some doubts when picking this name. The method both sets the state (if not set yet) and verifies it (if set). I'll change it to `setSubscriptionType`. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -77,6 +84,13 @@
         "Subscription to topics, partitions and pattern are mutually exclusive";
 
+    private void confirmSubscriptionType(SubscriptionType type) {
+        if (this.subscriptionType.equals(SubscriptionType.NONE))
```
Another nit: you can use == to test enums. It reads a little nicer, but more a matter of preference.

vahidhashemian replied:
Another one where I was debating what to choose. I simply picked `equals` because it is a built-in method. I can switch to using `==` if you think that reads better.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -634,12 +634,12 @@ public boolean matches(ClientRequest request) {
         metadata.update(TestUtils.singletonCluster(TopicConstants.GROUP_METADATA_TOPIC_NAME, 2), time.milliseconds());
 
-        assertFalse(subscriptions.partitionAssignmentNeeded());
+        assertTrue(subscriptions.partitionAssignmentNeeded());
```
This change brings up an interesting point. If a regex changes, but it has no impact on the set of matching topics, should the consumer rebalance?

vahidhashemian replied:
One scenario I can think of is when a topic that matches the regex is altered (e.g. its number of partitions changes). Currently, this does not seem to trigger a rebalance; even though a rebalance is required. So, it sounds like a rebalance should be done in regex subscription to cover cases like this. I hope that makes sense.

hachikuji replied:
We handle this case in `ConsumerCoordinator` I think. We basically store a snapshot of the metadata and request a rebalance if anything has changed (including changes to the number of partitions). Now that I'm thinking about it, this is probably why we didn't set `needsPartitionAssignment` to true when the regex changed. We were probably waiting for the metadata refresh to see if we needed to rebalance or not. I'm actually wondering if all that we were really missing was the call to `this.metadata.requestUpdate()` which you added above.

vahidhashemian replied:
I actually tested that scenario with [a simple Java consumer](https://github.com/vahidhashemian/kafka-tests/blob/trunk/src/com/github/vahidhashemian/kafka/samples/ConsumerSubscriptionSemantics.java), subscribed to a regex, and then, in a loop, kept polling every few seconds. When I changed the number of partitions that matched the regex, the rebalance listener of the consumer was not called, and its assigned partitions did not change.

If I comment out the added line `this.needsPartitionAssignment = true;` in `SubscriptionState`, I don't see a rebalance occur and the assigned partitions of the consumer remain the same after a subsequent regex subscription that's supposed to change them.

hachikuji replied:
Here's the code that should get triggered when the metadata is refreshed: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L161. One definite problem is that we didn't request a metadata update when the pattern changed. That would have meant that we wouldn't rebalance until the next periodic metadata refresh (which occurs every 5 minutes by default). I think the question is whether that logic works correctly or if there is another problem.

vahidhashemian replied:
You are right about number of partition changes being caught. One problem I noticed with my sample consumer is there seems to be a slight delay for consumer assignment to be refreshed once consumer subscription changes. It looks like with the scenarios I checked just refreshing the metadata should be enough. I'll update the patch. Thanks for your feedback.

hachikuji commented:
@vahidhashemian Thanks for the patch! Would it be much trouble to add an integration test (e.g. in `PlaintextConsumerTest` to verify that a second call to `subscribe(Pattern)` works correctly?

vahidhashemian commented:
@hachikuji Sure. I can do that. I had opened [KAFKA-3897](https://issues.apache.org/jira/browse/KAFKA-3897) to improve regex subscription testing in a separate PR.

hachikuji commented:
That works for me. Another small thing: could we change the title of the PR to reflect the problem you found (i.e. metadata should be refreshed immediately upon changing the subscription regex).

vahidhashemian commented:
Will do that too. Thanks.

vahidhashemian commented:
@hachikuji Actually, the PR fixes two issues. Metadata refresh fixes one (item 2 in description), and the SubscriptionType added fixes the other (item 1 in description). Do you think the title you suggested still works?

hachikuji commented:
Good point. In that case, the current title probably works.

vahidhashemian commented:
@hachikuji I added an integration test that covers both reported issues. Please advise if you see an issue. Thanks.

vahidhashemian commented:
Weird, some tests are randomly failing ([first](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/4494/), [second](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/4499/), [last](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/4504/) try), and all of the failures are from `TestUtils.waitUntilTrue(...)`. Perhaps the default 5000 ms is not a long enough wait time? I increased the wait time on the pattern subscription tests on the third try and they all passed. None fails when I test locally.

I see some other builds, like [this](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/4497/) and [this](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/4487/), also failed similarly.

hachikuji commented:
@vahidhashemian LGTM overall. I ran the test a few times locally and couldn't reproduce the failure, but jenkins tends to be more erratic. The only thought that comes to mind is that we somehow don't have up-to-date metadata when the consumer joins the group the first time.

vahidhashemian commented:
@hachikuji Could you please elaborate? Do you see that as a side-effect of this update, or something that still hasn't been addressed? Thanks.

hachikuji commented:
@vahidhashemian I don't see an actual problem, I'm just speculating what could cause the test to timeout. If you can reproduce the failure and see the consumer logs, we should be able to tell what's going on.

vahidhashemian commented:
Aah! Thanks for clarifying. I'll take a look at the logs. One thing I noticed happening occasionally for me locally, when I was testing my sample Java consumer, was a delay of over 20 seconds between when onPartitionsRevoked() and onPartitionsAssigned() were called. Not sure if that's related or not.

ewencp commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -184,6 +184,48 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+      this.consumers.head.assignment() == subscriptions.asJava
+    }, s"Expected partitions ${subscriptions.asJava} but actually got ${this.consumers.head.assignment()}")
+
+    val pattern2 = Pattern.compile("f.*") // only 'foo' matches this
```
@vahidhashemian The patch looks good, I just want to call out one thing that this test doesn't actually cover. We have other tests that validate that when we add topics, after a metadata refresh the new topic is included in the subscription when it matches the regex. This test, by checking for the removal of some topics, already validates that when metadata is updated we'll see the subscription updated to reflect the new regex. However, the _cause_ of the metadata refresh may be either a) the resubscription or b) the normal metadata refresh interval. For (b), to make some other tests fast, the base class sets it to only 100ms. If we want this test to also validate that the resubscription forces an _immediate_ metadata refresh and the subscription change is reflected promptly, I think we'd need to use a consumer with the metadata refresh interval overridden (see a few other tests in here that use custom settings for an example), add another topic before the resubscription (the only thing that will validate metadata changes), make sure it is included in the new regex, and then validate as we are already.

I think this is just some minor tweaking of the test, so hopefully straightforward to validate. A comment explaining what we're trying to make sure is happening would also be helpful as @hachikuji and I just had to spend some time thinking through what one of the other tests was actually evaluating.

vahidhashemian replied:
@ewencp Thanks for your feedback and catching the missing test. I'll add the test and the comments as you suggested.

vahidhashemian commented:
@ewencp I updated the test to cover what you mentioned. Also added some comments for the three pattern subscription tests. I realized that unsubscription is actually being tested in all three, so perhaps `testPatternUnsubscription()` is redundant.

ewencp commented:
LGTM, thanks @vahidhashemian for working through all the comments and improving docs on the existing tests!

ijuma commented:
@ewencp, seems like we should double-commit this to `0.10.0`. What do you think?

ewencp commented:
@ijuma I'm open to it, but skeptical -- would you consider this a critical fix that _must_ go into 0.10.0.1? It's an unfortunate misbehavior, but seems like an edge case I wouldn't consider critical. Updating regex subscriptions dynamically is not a particularly broad use case and this was only found through testing of a KIP as I understand it, not someone hitting a production issue with the new consumer.

All that said, if you want to cherry-pick, I will not object :)

ijuma commented:
@ewencp It's probably not critical, but it's a clear bug fix, it seemed low risk and it could surprise users in production. I wouldn't call an edge case because it's so easy to trigger (it's not like it requires a complex sequence of operations) even if it may be a bit rare in practice.

Do you agree that it's low risk? Happy to cherry-pick, if so. Otherwise, better to leave it indeed.

ewencp commented:
@ijuma Pushed to 0.10.0, and updated JIRA fix version. We should probably figure out how to distinguish between critical, low-risk, and not, but in this case, agreed I'm not too worried about existing functionality or compatibility.

ijuma commented:
Thanks!

reftel commented:
PlaintextConsumerTest.testPatternUnsubscription and PlaintextConsumerTest.testSubsequentPatternSubscription fail for me on commit 87b3ce with JDK 1.8.0_31. The failure messages are:

`java.lang.AssertionError: Expected partitions [topic-0, topic-1, tblablac-0, tblablac-1] but actually got []`
and

`java.lang.AssertionError: Expected partitions [topic-0, topic-1, foo-0] but actually got []`

All other test cases pass.

vahidhashemian commented:
@magnusr Looking at the recent failed builds it seems the issue is intermittent where one or more pattern subscription tests are failing, but there is no one test that always fail. I assume all tests pass locally for you?

ewencp commented:
@magnusr Yeah, would be good to know if this is consistent and only this test fails. Because we basically have system integration tests as unit tests, they end up causing transient failures. We've noticed some other tests failing under Jenkins and many seem to be relying on the default timeout from `TestUtils.waitUntilTrue`, so we may just need to increase that to make it more reliable across a variety of test hardware.

reftel commented:
@vahidhashemian: This was a local run on my laptop. 804 test run, 2 tests failed. Only tried once, so I don't know if it's intermittent or not.
